content/recipes/authoring/kata-snippets/full-tests.md

```diff
@@ -5,6 +5,84 @@ prev: /recipes/authoring/kata-snippets/sample-tests/
+- **Random tests should be run after fixed tests.** Not all testing frameworks allow for easy ordering of tests, but fixed tests should run, and eventually fail, before random tests.
+- **Difficulty should be consistent between test runs.** When randomized ranges are very large, it becomes possible for some users to receive many small, easy inputs while other users receive the exact opposite. Make sure that your random tests are built in a way which minimizes the chances for such a situation to occur. If you want to test difficult inputs, split the test cases into a set of easy ones and a set of difficult ones, and test them separately.
+- **Make debugging of random tests easy.** Sometimes, when random tests fail, they are very difficult to debug because input is very large and cannot be easily printed. If necessary, split your random tests into two batches: one with small, debuggable inputs, and the other one with proper, large input values. Note that both parts should still contain all applicable scenarios.
+
+## Performance Tests
+
+Some kata require solutions to be fast, or to fall into some performance category. For example, the author may only wish to accept solutions completing in (sub-)linear time. Building such test suites is not an easy task!
+
+- **Performance tests can be implemented in terms of random tests.** Usually, it's not reqired to have separate sections for random tests and performance tests. Testing of performance can be driven by size of random inputs. But still, to make debugging easier, it can be worth in some cases to have a set of separate random tests with small inputs.
+- If the difficulty of a kata is roughly proportional to the size of the input, it's usually **better to have a few tests with large input rather than many with small inputs.** For example, 200 tests with huge numbers or arrays is usually better than 10,000 tests with small ones (but see remarks on size of inputs above).
+- **The difference between accepted and rejected solutions should be easy to spot.** Ideally, accepted solutions should complete well under the time limit, while rejected solutions should time out consistently. Otherwise, you risk that solutions with valid complexity characteristics will time out, and users will be frustrated looking for micro-optimizations. This approach usually calls for very large inputs, so be careful!
+- **Performance tests should be consistent between runs.** It should not happen that one and the same solution sometimes passes, and sometimes fails, depending on randomized inputs.
+- When you are going to reject solutions based on their performance, **make sure that what you measure is what you want**. It's possible that most of the available time will be consumed not by user solution, but, for example, by your input generators, or reference solution. Make sure that poor performance of your test suite does not affect the final decision whether the submitted solution is performant enough or not!
```

Comment:
I still find that part not precise enough. So adding more details again (but less than before x) ). And suggesting again the thing about the slow solution (but in a softer way), because it's a general consideration, so it has imo its place in this "general documentation about testing performances". ;)

```suggestion
- **Make sure that what you measure is what you want**, when solutions are to be rejected based on their performances. With huge inputs, the random generation may be more consuming than the computation of the expected result. Hence the overall timing indication in the output panel is generally useless to ensure that the performances tests are actually discriminating the different kind of solutions as expected. The evaluation of the time actually used by the user's solution should be done (and compared to the reference solution if any) excluding the input generatin time.
- When it comes to maintainning a kata with performances requirements, it can be especially useful to have a solution whose the time complexity is supposed to be rejected on hand. Storing it, properly commented, in the test suite is then a very useful feature.
```

Reply:
```suggestion
- **Make sure that what you measure is what you want**, when solutions are to be rejected based on their performance. With huge inputs, the random generation may be more time-consuming than the computation of the expected result. Hence the overall timing indication in the output panel is generally useless to ensure that the performance tests are actually discriminating the different kinds of solutions as expected. The evaluation of the time actually used by the user's solution should be done (and compared to the reference solution if any) excluding the input generation time.
```